src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,49 +1,137 @@
+            elementArray[index] = value;
+            System.arraycopy(tempArray, 0, elementArray, index + 1, tempArray.length);
+        } else {
+            elementArray[index] = value;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            if (size + list.size() >= elementArray.length) {
+                grow();
+            }
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index > size - 1 || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Index is out of bounds");
+        }
+        return (T) elementArray[index];
     }
 
     @Override
     public void set(T value, int index) {
+        if (index > size - 1 || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Index is out of bounds");
+        }
+        elementArray[index] = value;
+    }
 
+    @Override
+    public T remove(T element) {
+        int newSize;
+        Object oldValue;
+        if ((newSize = size - 1) >= getIndexOfElement(element)) {
+            for (int i = 0; i < elementArray.length; i++) {
+                if (element == null && elementArray[i] == null) {
+                    System.arraycopy(elementArray, i + 1, elementArray, i, newSize);
+                    size = newSize;
+                    return null;
+                } else if (element.equals(elementArray[i])) {
+                    oldValue = elementArray[i];
+                    System.arraycopy(elementArray, i + 1, elementArray, i, newSize);
+                    size = newSize;
+                    return (T) oldValue;
+                }
+            }
+        }
+        throw new NoSuchElementException("There is no such element");
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        rangeCheck(index);
+        int newSize;
+        Object oldValue = elementArray[index];
+        if ((newSize = size - 1) > index || index > size - 1) {
+            System.arraycopy(elementArray, index + 1, elementArray, index, newSize);
+        }
+        size = newSize;
+        return (T) oldValue;
     }
 
-    @Override
-    public T remove(T element) {
-        return null;
+    public int getIndexOfElement(T element) {
+        for (int i = 0; i < elementArray.length; i++) {
+            if (element == null && elementArray[i] == null) {
+                return i;
+            } else if (element.equals(elementArray[i])) {
+                return i;
+            }
+        }
+        return -1;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void rangeCheck(int index) throws ArrayListIndexOutOfBoundsException {
+        if (index > size || index < 0) {
```

Comment:
```suggestion
        if (index >= size || index < 0) {
```
this way you don't need additional if checks in methods. This check won't work only for add by index, you can write custom check there